### ISSUE TITLE
Define own WatchErrorHandler for toolscache.Reflector

### ISSUE THREAD

**Issue by alexanderstephan** [kubernetes-sigs/controller-runtime#2987]
`client-go` sets a default handler for watching errors `func DefaultWatchErrorHandler(r *Reflector, err error)`, see https://github.com/kubernetes/client-go/blob/23900f492969db045e8a36396852381fd189569b/tools/cache/reflector.go#L154. Right now, if it is included in the `controller-runtime` it will use its logger `klog` by default. This leads to inconsistent log output by default. Therefore, I propose that the `controller-runtime` should overwrite this method, i.e., use its own handler.

This handler could look something like this:

```go
func WatchErrorHandler(r *toolscache.Reflector, err error) {
	log := ctrl.Log.WithValues("name", r.Name(), "description", r.TypeDescription())
	switch {
	case apierrors.IsResourceExpired(err) || apierrors.IsGone(err):
		// Don't set LastSyncResourceVersionUnavailable - LIST call with ResourceVersion=RV already
		// has a semantic that it returns data at least as fresh as provided RV.
		// So first try to LIST with setting RV to resource version of last observed object.
		log.Error(err, "watch closed")
	case err == io.EOF:
		// watch closed normally
	case err == io.ErrUnexpectedEOF:
		log.Error(err, "watch closed with unexpected EOF")
	default:
		utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.Name(), r.TypeDescription(), err))
	}
}
```
As you can see, we would use a logger with two fields of the reflector now. I exposed these fields recently (see https://github.com/kubernetes/client-go/commit/ce42c297f4b4444f810bedb6fcf10423eebafa5f), so the `controller-runtime` can essentially redefine the default handler of `client-go`, as the default handler also logs these fields.

**I would like to implement this, if it gets your approval.**

**Comment by sbueringer** [comment]
If I see correctly this part is currently refactored upstream: https://github.com/kubernetes/kubernetes/pull/127709

(cc @alvaroaleman)

**Comment by troy0820** [comment]
I see your point.  When implementing this [PR](https://github.com/kubernetes-sigs/controller-runtime/pull/2494/files) to be able to override the WatchErrorHandler it didn't occur to me that the klog would get in the way of the controller-runtime's logging, etc.  Implementing this seems doable and allowing this to be set to the default we provide or the overridden function in the cache options.  

However, we run into the problem of trying to stay within parity with upstream not using the upstream default handler.  Maintaining that seems like a lot of drift but still doable.

**Comment by alexanderstephan** [comment]
Thanks, so my take-away is that this seems reasonable to implement.

@sbueringer So, I should wait until the refactoring is over to avoid merge conflicts? Or, maybe start and use the branch of the PR as basis for now?

**Comment by sbueringer** [comment]
I could be absolutely misunderstanding this, but I was wondering if the problem still exists if we use the error handler that takes a context (and the logger out of the context).

I would recommend waiting until the PR in upstream is merged and we bumped to a Kubernetes version that has the change. Then we can check if we can simply switch to the error handler that takes a context.

**Comment by alexanderstephan** [comment]
Yes, okay seems good, let's wait then. Using the context here would also make sense to me. Also, we don't have this drifting problem that was mentioned before, from my understanding.

**Comment by alexanderstephan** [comment]
@sbueringer Do you have any updates on this whether this issue is still relevant now that I can see that the other PR is merged?

**Comment by sbueringer** [comment]
The new code is here: https://github.com/kubernetes/kubernetes/blob/0ddab1694579d523e3865c75ca44d6cdf1b0ef93/staging/src/k8s.io/client-go/tools/cache/reflector.go#L188-L201

Looks like they simply use the logger from the context. I assume we should be fine, but I would suggest let's bump to the alpha.1 next week and just try it out

**Comment by alexanderstephan** [comment]
Okay, yes, that looks good. Considering the context is initialized correctly now in the controller-runtime, I would assume that this issue can be closed. I suspect quite some changes here. Did you have a chance to try it out?

**Comment by sbueringer** [comment]
Added
```go
return nil, fmt.Errorf("abc")
```
here: https://github.com/kubernetes-sigs/controller-runtime/blob/85185e49ef2fe8345ba48c7181586ca66de8aaa3/pkg/cache/internal/informers.go#L363

Then ran `cache_suite_test.go`

I'm getting these logs
```
•E0308 12:22:26.615253   59566 reflector.go:200] "Failed to watch" err="failed to list *v1.Service: abc" logger="UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.0-alpha.3/tools/cache/reflector.go:285" type="*v1.Service"
E0308 12:22:48.688478   59566 reflector.go:200] "Failed to watch" err="failed to list *v1.Service: abc" logger="UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.0-alpha.3/tools/cache/reflector.go:285" type="*v1.Service"
```

So seems all good to me

/close

Feel free to reopen if you can still observe the issue with the CR main branch.

**Comment by sbueringer** [comment]
/reopen

Missed something ..

**Comment by sbueringer** [comment]
Okay I think with this PR It works as expected. Same code change as described in https://github.com/kubernetes-sigs/controller-runtime/issues/2987#issuecomment-2708219793

This time testing a controller configured with JSON logging

```json
{"ts":1741434916680.3196,"logger":"controller-runtime.cache.UnhandledError","caller":"runtime/runtime.go:226","msg":"Failed to watch","reflector":"pkg/mod/k8s.io/client-go@v0.33.0-alpha.3/tools/cache/reflector.go:285","type":"*v1alpha1.ExtensionConfig","err":"failed to list *v1alpha1.ExtensionConfig: abc"}
```

Another log for comparison

```json
{"ts":1741434890289.1465,"logger":"controller-runtime.metrics","caller":"server/server.go:247","msg":"Serving metrics server","v":0,"bindAddress":":8443","secure":true}
```

**Comment by alexanderstephan** [comment]
Yes, this PR looks like what I would have expected. Very nice that this is now nearing completion. Thanks so much for your support!
